From 7f1fc8b3852647ce4d782366253027f5dce12fc8 Mon Sep 17 00:00:00 2001
From: Ben <zyhfish@163.com>
Date: Thu, 24 May 2018 08:42:58 +0800
Subject: [PATCH] DNN-20580: import pages in order.

---
 .../DnnExportImport/Components/Services/PagesExportService.cs    | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/DNN Platform/Modules/DnnExportImport/Components/Services/PagesExportService.cs b/DNN Platform/Modules/DnnExportImport/Components/Services/PagesExportService.cs
index f36dd696ca..f4aee3736b 100644
--- a/DNN Platform/Modules/DnnExportImport/Components/Services/PagesExportService.cs	
+++ b/DNN Platform/Modules/DnnExportImport/Components/Services/PagesExportService.cs	
@@ -143,16 +143,19 @@ private void ProcessImportPages()
 
             var localTabs = _tabController.GetTabsByPortal(portalId).Values.ToList();
 
-            var exportedTabs = Repository.GetItems<ExportTab>(x => x.IsSystem == (Category == Constants.Category_Templates)).ToList(); // ordered by TabID
+            var exportedTabs = Repository.GetItems<ExportTab>(x => x.IsSystem == (Category == Constants.Category_Templates))
+                .OrderBy(t => t.Level).ThenBy(t => t.ParentId).ThenBy(t => t.TabOrder).ToList(); 
+
             //Update the total items count in the check points. This should be updated only once.
             CheckPoint.TotalItems = CheckPoint.TotalItems <= 0 ? exportedTabs.Count : CheckPoint.TotalItems;
             if (CheckPointStageCallback(this)) return;
             var progressStep = 100.0 / exportedTabs.OrderByDescending(x => x.Id).Count(x => x.Id < _totals.LastProcessedId);
 
+            var index = 0;
             foreach (var otherTab in exportedTabs)
             {
                 if (CheckCancelled(_exportImportJob)) break;
-                if (_totals.LastProcessedId > otherTab.Id) continue; // this is the exported DB row ID; not the TabID
+                if (_totals.LastProcessedId > index) continue; // this is the exported DB row ID; not the TabID
 
                 ProcessImportPage(otherTab, exportedTabs, localTabs);
 
@@ -160,7 +163,7 @@ private void ProcessImportPages()
                 CheckPoint.Progress += progressStep;
                 if (CheckPointStageCallback(this)) break;
 
-                _totals.LastProcessedId = otherTab.Id;
+                _totals.LastProcessedId = index++;
                 CheckPoint.StageData = JsonConvert.SerializeObject(_totals);
             }
 
